src/replay/consensus/core.zig

```diff
@@ -2205,4 +2702,1018 @@ test "computeBankStats - same weight selects lower slot" {
+    var prng = std.Random.DefaultPrng.init(43);
+    const random = prng.random();
+
+    const root = SlotAndHash{ .slot = 0, .hash = Hash.initRandom(random) };
+    var fixture = try TestFixture.init(allocator, root);
+    defer fixture.deinit(allocator);
+
+    fixture.slot_tracker.get(0).?.state.hash.set(root.hash);
+
+    var replay_tower = try createTestReplayTower(1, 0.67);
+    defer replay_tower.deinit(allocator);
+
+    const auth_voter_kp = sig.identity.KeyPair.generate();
+    const vote_account_pubkey = Pubkey.initRandom(random);
+
+    const epoch_tracker: EpochTracker = .{ .schedule = .DEFAULT, .epochs = .empty };
+    const account_reader: AccountReader = .noop;
+
+    const result = try generateVoteTx(
+        allocator,
+        vote_account_pubkey,
+        &.{auth_voter_kp},
+        null,
+        .same_fork,
+        &replay_tower,
+        account_reader,
+        &fixture.slot_tracker,
+        &epoch_tracker,
+    );
+
+    try testing.expectEqual(.non_voting, result);
+}
+
+test "generateVoteTx - no last voted slot returns failed" {
+    const allocator = testing.allocator;
+    var prng = std.Random.DefaultPrng.init(44);
+    const random = prng.random();
+
+    const root = SlotAndHash{ .slot = 0, .hash = Hash.initRandom(random) };
+    var fixture = try TestFixture.init(allocator, root);
+    defer fixture.deinit(allocator);
+
+    fixture.slot_tracker.get(0).?.state.hash.set(root.hash);
+
+    var replay_tower = try createTestReplayTower(1, 0.67);
+    defer replay_tower.deinit(allocator);
+    replay_tower.last_vote = .{ .vote = .{ .slots = &.{}, .hash = Hash.ZEROES, .timestamp = null } };
+
+    const node_kp = sig.identity.KeyPair.generate();
+    const auth_voter_kp = sig.identity.KeyPair.generate();
+    const vote_account_pubkey = Pubkey.initRandom(random);
+
+    const epoch_tracker: EpochTracker = .{ .schedule = .DEFAULT, .epochs = .empty };
+    const account_reader: AccountReader = .noop;
+
+    const result = try generateVoteTx(
+        allocator,
+        vote_account_pubkey,
+        &.{auth_voter_kp},
+        node_kp,
+        .same_fork,
+        &replay_tower,
+        account_reader,
+        &fixture.slot_tracker,
+        &epoch_tracker,
+    );
+
+    try testing.expectEqual(.failed, result);
+}
+
+test "generateVoteTx - slot not in tracker returns failed" {
+    const allocator = testing.allocator;
+    var prng = std.Random.DefaultPrng.init(45);
+    const random = prng.random();
+
+    const root = SlotAndHash{ .slot = 0, .hash = Hash.initRandom(random) };
+    var fixture = try TestFixture.init(allocator, root);
+    defer fixture.deinit(allocator);
+
+    fixture.slot_tracker.get(0).?.state.hash.set(root.hash);
+
+    var replay_tower = try createTestReplayTower(1, 0.67);
+    defer replay_tower.deinit(allocator);
+
+    replay_tower.last_vote = sig.consensus.vote_transaction.VoteTransaction{
+        .tower_sync = sig.runtime.program.vote.state.TowerSync{
```

Comment:
Any reason for these type specifications? These could both be written as `.{`. Same comment in general for all instances of this.